### PR TITLE
BAN-2033 Add Apple news 'Videos' export feature

### PR DESCRIPTION
**Pull request by damiendillon**
Use the Apple News Component `EmbedWebVideo` for YouTube, Dailymotion and Vimeo videos.

Also add facebook_post (for Facebook videos) & tweet components.

https://mydevex.atlassian.net/browse/BAN-2033

### CODE REVIEW

File: lib/article_json/export/apple_news/elements/embed.rb [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+          def build_facebook_video_url
+            embed_params = embed_id.to_s.split("/")
+            username = embed_params[0]
+            embed_id = embed_params[1]
+            "https://www.facebook.com/#{username}/videos/#{embed_id}"
+          end
+
+          def build_twitter_url
+            embed_params = embed_id.to_s.split("/")
```
Comment (ozgun) [inline lib/article_json/export/apple_news/elements/embed.rb]: Not a big deal, but could be simplified like this:
```ruby
username, embed_id = embed_id.to_s.split("/").take(2)
```

File: lib/article_json/export/apple_news/elements/embed.rb [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+          def role
+            @role ||=
+              case embed_type
+              when :youtube_video, :vimeo_video, :dailymotion_video
+                :embedwebvideo
+              when :facebook_video
+                :facebook_post
+              when :tweet
+                :tweet
+              when :slideshare
```
Comment (ozgun) [inline lib/article_json/export/apple_news/elements/embed.rb:27]: Do we need `:slideshare` and `:soundcloud` cases here? Or does it mean they will be coming soon?

File: lib/article_json/export/apple_news/exporter.rb [hunk 3]
```diff
@@ -23,6 +23,8 @@ def components
               ArticleJSON::Export::AppleNews::Elements::Base
                 .build(element)
                 &.export
+            end.reject do |hash|
```
Comment (ozgun) [inline lib/article_json/export/apple_news/exporter.rb]: How about `end.reject { |hash| hash&.empty? }`

**Comment by ozgun** [comment]
I have a couple of minor suggestions, but they don't seem to be blockers.

File: lib/article_json/export/apple_news/elements/embed.rb (hunk 1 above)
Comment (damiendillon) [reply lib/article_json/export/apple_news/elements/embed.rb]: image

File: lib/article_json/export/apple_news/elements/embed.rb (hunk 2 above)
Comment (damiendillon) [reply lib/article_json/export/apple_news/elements/embed.rb:27]: It does not necessarily mean that they are coming soon but they are elements that we send from article_json and they act as placeholders, showing that they have not been dealt with.

File: lib/article_json/export/apple_news/exporter.rb (hunk 3 above)
Comment (damiendillon) [reply lib/article_json/export/apple_news/exporter.rb]: Amened

File: lib/article_json/export/apple_news/elements/embed.rb (hunk 1 above)
Comment (damiendillon) [reply lib/article_json/export/apple_news/elements/embed.rb]: The suggestion should work in theory, but the variable cannot be reused it seems.

Amened without the take `method`.

File: lib/article_json/export/apple_news/exporter.rb (hunk 3 above)
Comment (damiendillon) [reply lib/article_json/export/apple_news/exporter.rb]: We actually have nil values too that we want to remove.

Comment (damiendillon) [reply lib/article_json/export/apple_news/exporter.rb]: Made it a one liner.

**Comment by damiendillon** [comment]
> I have a couple of minor suggestions, but they don't seem to be blockers.

If you are happy with it..

**Comment by ozgun** [comment]
@damiendillon Thanks for the changes :)

**Comment by damiendillon** [comment]
> @damiendillon Thanks for the changes :)

I know there is no push pull on here but if you approve, I think leaving a 'plus one' is clearer for whoever clicks for the merge.

**Comment by ozgun** [comment]
Sure! +1

**Comment by EdLittle** [comment]
+1 LGTM

**Comment by dsager** [comment]
+1 GJ!

File: lib/article_json/export/apple_news/elements/embed.rb [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+          def build_embeded_vimeo_url
+            "https://player.vimeo.com/video/#{embed_id}"
+          end
+
+          def build_embeded_dailymotion_url
+            "https://geo.dailymotion.com/player.html?video=#{embed_id}"
+          end
+
+          def build_facebook_video_url
+            username, id = embed_id.to_s.split("/")
```
Comment (dsager) [inline lib/article_json/export/apple_news/elements/embed.rb:76]: you can also tell into how many items you want to split the string: `split("/", 2)`

Comment (damiendillon) [reply lib/article_json/export/apple_news/elements/embed.rb:76]: Ah, good to know. I can add that.

Comment (ozgun) [reply lib/article_json/export/apple_news/elements/embed.rb:76]: Nice, I didn't know about the second parameter!

**Pull request merged**